### PR TITLE
[ACR] az acr import: Support no-wait 

### PR DESCRIPTION
**Pull request by CarolineNB**
**Description**

Fixes https://github.com/Azure/azure-cli/issues/17851

**Testing Guide**
`az acr import -n MyRegistry --source sourceregistry.azurecr.io/sourcerepository:sourcetag --no-wait`

---

This checklist is used to make sure that common guidelines for a pull request are followed.

- [ ] The PR title and description has followed the guideline in [Submitting Pull Requests](https://github.com/Azure/azure-cli/tree/dev/doc/authoring_command_modules#submitting-pull-requests).

- [ ] I adhere to the [Command Guidelines](https://github.com/Azure/azure-cli/blob/dev/doc/command_guidelines.md).

- [ ] I adhere to the [Error Handling Guidelines](https://github.com/Azure/azure-cli/blob/dev/doc/error_handling_guidelines.md).

### CODE REVIEW

**Comment by yonzhan** [comment]
ACR

File: src/azure-cli/azure/cli/command_modules/acr/_help.py [hunk 1]
```diff
@@ -264,6 +264,9 @@
     text: |
         az acr import -n MyRegistry --source sourcerepository:sourcetag -t targetrepository:targettag \\
             -r /subscriptions/00000000-0000-0000-0000-000000000000/resourceGroups/sourceResourceGroup/providers/Microsoft.ContainerRegistry/registries/sourceRegistry
+  - name: Import an image and return right after starting an import.
```
Comment (adewaleo) [inline src/azure-cli/azure/cli/command_modules/acr/_help.py:267]: Maybe reword in a way like....
```suggestion
  - name: Import an image without waiting for completion.
```
or 
```suggestion
  - name: Import an image. Do not wait for completion.
```
Just some suggestions to consider.

Comment (adewaleo) [reply src/azure-cli/azure/cli/command_modules/acr/_help.py:267]: Also you can update the short summary above to add:
```
For large images consider using `--no-wait`.
```

File: src/azure-cli/azure/cli/command_modules/acr/import.py [hunk 2]
```diff
@@ -94,18 +95,23 @@ def acr_import(cmd,
...
+        return result_poller
+
     return _handle_result(cmd, result_poller, source_registry, source_image, registry)
 
 
 def _handle_result(cmd, result_poller, source_registry, source_image, registry):
     from msrestazure.azure_exceptions import ClientException
     try:
-        result = LongRunningOperation(cmd.cli_ctx, 'Importing image...')(result_poller)
+        result = LongRunningOperation(cmd.cli_ctx, 'Importing image...', 'Import has finished')(result_poller)
```
Comment (adewaleo) [inline src/azure-cli/azure/cli/command_modules/acr/import.py:113]: maybe consider:
```suggestion
        result = LongRunningOperation(cmd.cli_ctx, 'Importing image...', 'Import succeeded')(result_poller)
```
or something else like "Import succeeded" or "Import completed" or just "success".

File: src/azure-cli/azure/cli/command_modules/acr/import.py [hunk 3]
```diff
@@ -94,18 +95,23 @@ def acr_import(cmd,
...
                                               untagged_target_repositories=repository,
                                               mode=ImportMode.force.value if force else ImportMode.no_force.value)
+
     result_poller = client.import_image(
         resource_group_name=resource_group_name,
         registry_name=registry_name,
         parameters=import_parameters)
 
+    if no_wait:
+        return result_poller
```
Comment (adewaleo) [inline src/azure-cli/azure/cli/command_modules/acr/import.py:105]: Simply returning the poller might not be the best way to go about this. 

The [authoring command modules](https://github.com/Azure/azure-cli/blob/dev/doc/authoring_command_modules/authoring_commands.md) section on supporting `--no-wait`

Mentions the cli core method `azure.cli.core.util.sdk_no_wait` defined here in [util.py](https://github.com/Azure/azure-cli/blob/dev/src/azure-cli-core/azure/cli/core/util.py#L646).

```python
sdk_no_wait(no_wait, func, *args, **kwargs)
```
This method does calls the function you pass with the supplied args and kwargs, and sets/updates the arguments to include `polling: False`. 

Given our code doesn't rely on sdk_no_wait, some approaches come to mind on how we can either use it or copy the behavior of sdk_no_wait when handling result.

File: src/azure-cli/azure/cli/command_modules/acr/import.py [hunk 4]
```diff
@@ -94,18 +95,23 @@ def acr_import(cmd,
...
+
     return _handle_result(cmd, result_poller, source_registry, source_image, registry)
 
 
 def _handle_result(cmd, result_poller, source_registry, source_image, registry):
     from msrestazure.azure_exceptions import ClientException
     try:
-        result = LongRunningOperation(cmd.cli_ctx, 'Importing image...')(result_poller)
```
Comment (adewaleo) [inline src/azure-cli/azure/cli/command_modules/acr/import.py:108]: Inline with my comment above, we might need to rethink how we approach handle result when we have a wait versus a no wait. There are a few examples of calls to sdk_no_wait here in this [file](https://github.com/Azure/azure-cli/blob/dev/src/azure-cli/azure/cli/command_modules/vm/custom.py#L1013).

File: src/azure-cli/azure/cli/command_modules/acr/import.py [hunk 5]
```diff
@@ -94,18 +95,23 @@ def acr_import(cmd,
...
+        result = LongRunningOperation(cmd.cli_ctx, 'Importing image...', 'Import has finished')(result_poller)
+
     except CLIError as e:
         try:
             # if command fails, it might be because user specified registry twice in --source and --registry
```
Comment (adewaleo) [inline src/azure-cli/azure/cli/command_modules/acr/import.py:117]: finally let's make sure that with or without no-wait (if user enters input incorrectly) and service rejects the request in the event of no-wait, we show the user the warning below..

**Review by adewaleo — CHANGES_REQUESTED** [review]
Nice work @CarolineNB. some comments and feedback.

File: src/azure-cli/azure/cli/command_modules/acr/import.py [hunk 6]
```diff
@@ -35,8 +35,9 @@ def acr_import(cmd,
                target_tags=None,
                resource_group_name=None,
                repository=None,
-               force=False):
-
+               force=False,
+               no_wait=False,
+               ):
```
Comment (jaysterp) [inline src/azure-cli/azure/cli/command_modules/acr/import.py:40]: @CarolineNB small style comment: please remove the comma after the last parameter. Also `):` does not need to be on a new line.

**Review by Juliehzl — COMMENTED** [review]
1. Please add test for --no-wait to make sure it work
2. Then recommend to add wait command to query

File: src/azure-cli/azure/cli/command_modules/acr/import.py [hunk 7]
```diff
@@ -95,45 +96,44 @@ def acr_import(cmd,
...
-        return result_poller
+            return LongRunningOperation(cmd.cli_ctx, 'Importing image...')(result_poller)
 
-    return _handle_result(cmd, result_poller, source_registry, source_image, registry)
+    except CLIError as e:
+        _handle_result(e, cmd, source_registry, source_image, registry)
 
 
-def _handle_result(cmd, result_poller, source_registry, source_image, registry):
+def _handle_result(e, cmd, source_registry, source_image, registry):
```
Comment (jaysterp) [inline src/azure-cli/azure/cli/command_modules/acr/import.py:115]: @CarolineNB nit: since this class is now an exception handler only, we can rename it to `_handle_exception`.

Comment (adewaleo) [reply src/azure-cli/azure/cli/command_modules/acr/import.py]: agreed. Maybe name it `_handle_import_exception`

File: src/azure-cli/azure/cli/command_modules/acr/tests/latest/test_acr_commands.py [hunk 8]
```diff
@@ -245,6 +245,36 @@ def test_acr_create_replication(self, resource_group, resource_group_location):
         # test acr delete
         self.cmd('acr delete -n {registry_name} -g {rg} -y')
 
+    @ResourceGroupPreparer()
+    def test_acr_import_no_wait(self, resource_group):
+        source_registry_name = self.create_random_name("sourceregistrysamesub", 40)
+        registry_name = self.create_random_name("targetregistry", 20)
```
Comment (adewaleo) [inline src/azure-cli/azure/cli/command_modules/acr/tests/latest/test_acr_commands.py:251]: registry_name isn't being used
```suggestion
```

File: src/azure-cli/azure/cli/command_modules/acr/_help.py [hunk 9]
```diff
@@ -264,7 +264,7 @@
     text: |
         az acr import -n MyRegistry --source sourcerepository:sourcetag -t targetrepository:targettag \\
             -r /subscriptions/00000000-0000-0000-0000-000000000000/resourceGroups/sourceResourceGroup/providers/Microsoft.ContainerRegistry/registries/sourceRegistry
-  - name: Import an image and return right after starting an import.
+  - name: Import an image without waiting for completion.
```
Comment (adewaleo) [inline src/azure-cli/azure/cli/command_modules/acr/_help.py:267]: @jaysterp it might be overkill, but I wonder if we should hint that this doesn't check for success.
maybe say: 
```suggestion
  - name: Import an image without waiting for successful completion.
```

File: src/azure-cli/azure/cli/command_modules/acr/tests/latest/recordings/test_check_name_availability.yaml [hunk 10]
```diff
@@ -17,8 +17,8 @@ interactions:
       ParameterSetName:
       - -n
       User-Agent:
-      - python/3.7.9 (Windows-10-10.0.19041-SP0) msrest/0.6.18 msrest_azure/0.6.3
-        azure-mgmt-containerregistry/3.0.0rc16 Azure-SDK-For-Python AZURECLI/2.16.0
+      - python/3.8.10 (Windows-10-10.0.19041-SP0) msrest/0.6.21 msrest_azure/0.6.3
```
Comment (adewaleo) [inline src/azure-cli/azure/cli/command_modules/acr/tests/latest/recordings/test_check_name_availability.yaml:20]: is checking in these new recordings necessary? If not, it's okay, but also consider removing unnecessary check-ins to keep PR concise.

**Review by adewaleo — COMMENTED** [review]
Overall looks good to me. However, left some suggestions and comments.

File: src/azure-cli/azure/cli/command_modules/acr/import.py [hunk 11]
```diff
@@ -35,7 +36,8 @@ def acr_import(cmd,
                target_tags=None,
                resource_group_name=None,
                repository=None,
-               force=False):
+               force=False,
+               no_wait=False):
```
Comment (jaysterp) [inline src/azure-cli/azure/cli/command_modules/acr/import.py:40]: @CarolineNB please add the comment `# pylint: disable=too-many-locals` to suppress the style warning "Too many local variables". Since these variables are command parameters, we can ignore this error. Please reference [acr_task_create](https://github.com/Azure/azure-cli/blob/aaffbc5cafa4535aadc180136ec094f1725a0f38/src/azure-cli/azure/cli/command_modules/acr/task.py#L43).

File: src/azure-cli/azure/cli/command_modules/acr/import.py [hunk 12]
```diff
@@ -94,41 +96,44 @@ def acr_import(cmd,
...
-    result_poller = client.import_image(
-        resource_group_name=resource_group_name,
-        registry_name=registry_name,
-        parameters=import_parameters)
 
-    return _handle_result(cmd, result_poller, source_registry, source_image, registry)
+    try:
+        if no_wait:
+            return sdk_no_wait(no_wait, client.import_image, resource_group_name, registry_name, import_parameters)
+        else:
```
Comment (jaysterp) [inline src/azure-cli/azure/cli/command_modules/acr/import.py:103]: @CarolineNB pylint is flagging this if/else with the "Unnecessary "else" after "return"" style error. You can see how to fix this by referencing [no-else-return](https://pheanex.github.io/pylint/#r1705-no-else-return).

Pylint error: https://dev.azure.com/azure-sdk/public/_build/results?buildId=973250&view=logs&jobId=36dd4138-4d53-5e46-00d9-e5cd9744be05&j=36dd4138-4d53-5e46-00d9-e5cd9744be05&t=1cf3879c-0a42-5ccd-7693-7a4781d739d8

File: src/azure-cli/azure/cli/command_modules/acr/_help.py (hunk 9 above)
[the remark this answers is not in the document]
Comment (jaysterp) [reply src/azure-cli/azure/cli/command_modules/acr/_help.py:267]: @adewaleo I like this idea. Since imageImport does not provide a GET API that users could leverage to check their import status, it is a good user experience to warn them of this if they opt out of waiting. We can recommend them in the warning to run `az acr repository show-tags` to confirm the import succeeded (however this does not cover the case where import overwrites tags). @nathana1 what do you think?

`--no-wait=True` will still catch failures in the front end of the request after it lands in our RP and before we send the request to the backend for processing. These errors (request validation, etc) are all thrown before we create the async response. Failure in the backend to actually perform the import will not be surfaced.

Comment (jaysterp) [reply src/azure-cli/azure/cli/command_modules/acr/_help.py:267]: @adewaleo I agree with this suggestion. @CarolineNB

[the remark this answers is not in the document]
Comment (nathana1) [reply src/azure-cli/azure/cli/command_modules/acr/_help.py:267]: I think adding a warning to the description/documentation would be a good idea.  I don't think we'd want to add a warning to every time the command is run with this flag.

[the remark this answers is not in the document]
Comment (adewaleo) [reply src/azure-cli/azure/cli/command_modules/acr/_help.py:267]: yeah I was thinking more along the line of documentation warning. (I do think logging a warning on `--no-wait` imports could also be helpful)

**Comment by iamjasonp** [comment]
Rebase to dev completed with 2nd pair of eyes cc: @jaysterp

**Review by jaysterp — APPROVED** [review]

**Review by adewaleo — APPROVED** [review]

**Comment by jaysterp** [comment]
@Juliehzl we are ready to merge. Please take a look. Thanks!

**Review by Juliehzl — APPROVED** [review]
